Sprint 17

Peppa-Akk opened the pull request:
Sprint 17:
- Add context menu (pin/unpin, edit, delete);
- Add filters;
- Add screenshot test;
- Add AppMetrica;
- Add statistic flow;
- Localise app to russian and english;
- Realise search controller.

Peppa-Akk committed:
Update UI:
- Add scroll; - Add CollectionView; - Configure CollectionView; - Add new view (Schedule); - Configure Schedule View.  Start with CoreData: - Add to project data model (.cdatamodeld file); - Configure data model (entity and relationship); - Init Persistent Container. 

Peppa-Akk committed:
Add subtitle for schedule button

Peppa-Akk committed:
update CoreDataStores:
- add new tracker to Core Data;
- apply category relationship for tracker;
- add new record to Core Data;
- delete record (with id and date);

Fix date logic (now use start day);
Fix present Event View.

Peppa-Akk committed:
sprint 16:
- Add logic of category;
- Implemented CoreData to category;
- Category logic is written on the MVVM;
- Add category to Trackers;
- Add Onboard View;
- Add Keychain property to check for the first entrance in application.

Peppa-Akk committed:
Sprint 16:
- Fix button (now it's not at the OnboardingVIewController);
- FIx selection of category.

Peppa-Akk committed:
Sprint 17:
- Add context menu (pin/unpin, edit, delete);
- Add filters;
- Add screenshot test;
- Add AppMetrica;
- Add statistic flow;
- Localise app to russian and english;
- Realise search controller.

helecta commented on Podfile.lock:
```diff
@@ -0,0 +1,112 @@
+PODS:
```
Обычно в гит не добавляют сами исходники подов, только главный pod файл с перечислением либ. Дальше каждый разработчик у себя скачивает по этому файлу нужные библиотеки. 
Вот [тут](https://stellarflavor.com/2021/12/25/how-to-remove-pushed-pods-folder-from-git-repository/) описано как можно удалить эту папку из репозиторияПо этой причине у тебя сейчас получился такой огромный пул-реквест

helecta commented on Tracker/CollectionOfTrackers/TrackerCollectionViewCell.swift:
```diff
@@ -188,12 +219,14 @@ extension TrackerCollectionViewCell {
         self.trackerID = tracker.id
         self.isCompletedToday = isCompletedToday
         self.indexPath = indexPath
+        self.isPin = tracker.isPin
```
Тут `self` необязательно использовать, поэтому лучше его убрать.
Подробнее можешь почитать в [код стайле](https://github.com/tutu-ru/swift-style-guide?tab=readme-ov-file#Комментарии)

helecta commented on Tracker/CollectionOfTrackers/TrackerCollectionViewCell.swift:
```diff
@@ -210,3 +243,37 @@ extension TrackerCollectionViewCell {
+            let fixAction = UIAction(
+                title: self.isPin ? NSLocalizedString("Unpin", comment: "") : NSLocalizedString("Pin", comment: "")
+            ) { _ in
+                guard let trackerID = self.trackerID else { return }
```
В замыкание для избежания retain cycl'ов, необходимо передавать слабую ссылку на `self`

helecta commented on Tracker/CollectionOfTrackers/TrackerCollectionViewCell.swift:
```diff
@@ -210,3 +243,37 @@ extension TrackerCollectionViewCell {
+extension TrackerCollectionViewCell: UIContextMenuInteractionDelegate {
+    
+    func contextMenuInteraction(_ interaction: UIContextMenuInteraction, configurationForMenuAtLocation location: CGPoint) -> UIContextMenuConfiguration? {
+        analyticService.report(event: "Long press on tracker's cell to open a context menu on TrackersViewController", params: ["event": "click", "screen": "Main", "item": "cell"])
```
Обычно самые частые проблемы с аналитикой - отправка неправильных параметров. Одна платформа с большой буквы шлет, другая с маленькой. И все, данные не сходятся. Удобней исправлять это, когда все они находятся в одном месте, а не раскиданы по всему приложению

Поэтому лучше было бы все значения параметров аналитики убрать в AnalyticsService.
Создать там метод, например:
```
    func didOpenMain() {
        report(event: "open", params: ["screen": "Main"])
    }
```

И вызывать тут уже его `analyticsService.didOpenMain()`
Аналогично сделать для всех событий.

helecta commented on Tracker/CollectionOfTrackers/TrackerCollectionViewCell.swift:
```diff
@@ -210,3 +243,37 @@ extension TrackerCollectionViewCell {
+            }
+            
+            let editAction = UIAction(title: NSLocalizedString("Edit", comment: "")) { _ in
+                self.analyticService.report(event: "Chose edit option in tracker's context menu", params: ["event": "click", "screen": "Main", "item": "edit"])
```
необходимо передавать слабую ссылку на self

helecta commented on Tracker/CoreData/CoreDataStores/CategoryStore.swift:
```diff
@@ -0,0 +1,124 @@
+    weak var delegate: StoreDelegate?
+    
+    private let context: NSManagedObjectContext
+    static let shared = CategoryStore()
```
Предполагается, что CategoryStore будет использоваться в одном экземпляре, т.е. через shared. Тогда стоит добавить приватный init. Так мы реализуем паттерн  [singleton](https://www.hackingwithswift.com/example-code/language/what-is-a-singleton) .

helecta replied:
Не забывай организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

Подробнее почитать можно тут: [Организация кода](https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization)

helecta commented on Tracker/CoreData/CoreDataStores/CategoryStore.swift:
```diff
@@ -0,0 +1,124 @@
+    
+    convenience override init() {
+        let context = (UIApplication.shared.delegate as! AppDelegate).persistentContainer.viewContext
+        try! self.init(context: context)
```
здесь было бы хорошо обработать ошибку

helecta commented on Tracker/CoreData/CoreDataStores/CategoryStore.swift:
```diff
@@ -0,0 +1,124 @@
+        request.resultType = .managedObjectIDResultType
+        request.predicate = NSPredicate(format: "%K == %@", #keyPath(TrackerCategoryCoreData.categoryID), id.uuidString)
+
+        let categoryID = try! context.execute(request) as! NSAsynchronousFetchResult<NSFetchRequestResult>
```
здесь было бы хорошо обработать ошибку

helecta commented on Tracker/CoreData/CoreDataStores/CategoryStore.swift:
```diff
@@ -0,0 +1,124 @@
+        guard let categoryID = categoryCoreData.categoryID else { throw CategoryStoreError.decodingErrorInvalidID }
+        guard let title = categoryCoreData.title else { throw CategoryStoreError.decodingErrorInvalidTitle }
+        var trackers: [Tracker] = []
+        let allTrackers = categoryCoreData.trackers?.allObjects as! [TrackerCoreData]
```
Тут force cast: `as! [TrackerCoreData]`, стоит переписать без него. Т.к он может приводить к падению приложения, считается что падение приложения хуже, чем не рабочее состояние. Нерабочее состояние можно обработать, сохранить последнее состояние приложения, что бы позже пользователь мог продолжить работу с того места на котором остановился. С падением уже ничего сделать нельзя, обычно это вызывает негатив у пользователей.

helecta commented on Tracker/CoreData/CoreDataStores/CategoryStore.swift:
```diff
@@ -0,0 +1,124 @@
+        request.predicate = NSPredicate(format: "%K == %@", #keyPath(TrackerCategoryCoreData.categoryID), id.uuidString)
+
+        let categoryID = try! context.execute(request) as! NSAsynchronousFetchResult<NSFetchRequestResult>
+        let id = categoryID.finalResult![0] as! NSManagedObjectID
```
Тут и force cast и force unwrap, такие операции следует исбегать

helecta commented on Tracker/CoreData/CoreDataStores/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,150 @@
+//import CoreData
+//
+////MARK: - Errors
+//enum TrackerStoreCategoryError: Error {
```
Закомментированный код перед сдачей работы нужно удалятьЗакомментированный код перед сдачей работы нужно удалять

Peppa-Akk replied:
Удалял этот файл. У меня в проекте отсутствует

helecta commented on Tracker/CoreData/CoreDataStores/TrackerCoreData+.swift:
```diff
@@ -0,0 +1,31 @@
+        return NSFetchRequest<TrackerCoreData>(entityName: "TrackerCoreData")
+    }
+
+    @NSManaged public var trackerID: UUID?
```
Здесь можно без public. Обычно public используется при написании отдельно подключаемых модулей или либ. Т.е public помечают то, что будет доступно извне модуля. 
В этом проекте только один модуль, поэтому здесь он не обязателен.
Подробнее про [уровни доступа](https://habr.com/ru/articles/755744/)

Peppa-Akk replied:
Не нашел данный файл в проекте

helecta commented on Tracker/CoreData/CoreDataStores/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,120 @@
+        
+        let trackerRecordCoreData = TrackerRecordCoreData(context: context)
+        updateExistingRecord(trackerRecordCoreData, with: record, and: trackerID)
+        (UIApplication.shared.delegate as! AppDelegate).saveContext()
```
force cast

helecta commented on Tracker/CoreData/CoreDataStores/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,120 @@
+        recordCoreData.recordID = record.id
+        recordCoreData.date = record.date
+        trackerStore.relationshipWithRecord(recordCoreData, by: trackerID)
+        (UIApplication.shared.delegate as! AppDelegate).saveContext()
```
force cast

helecta commented on Tracker/CoreData/CoreDataStores/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,120 @@
+    func deleteRecord(with recordID: UUID, and date: Date) {
+        
+        context.delete(fetchRecord(with: fetchRecordID(with: recordID, and: date)))
+        (UIApplication.shared.delegate as! AppDelegate).saveContext()
```
force cast

helecta commented on Tracker/CoreData/CoreDataStores/TrackerStore.swift:
```diff
@@ -0,0 +1,180 @@
+    }
+    
+    convenience override init() {
+        let context = (UIApplication.shared.delegate as! AppDelegate).persistentContainer.viewContext
```
force cast

helecta commented on Tracker/CoreData/CoreDataStores/TrackerStore.swift:
```diff
@@ -0,0 +1,180 @@
+    func addNewTracker(_ tracker: Tracker, with categoryID: NSManagedObjectID) {
+        let trackerCoreData = TrackerCoreData(context: context)
+        updateExistingTracker(trackerCoreData, with: tracker, and: categoryID)
+        (UIApplication.shared.delegate as! AppDelegate).saveContext()
```
force cast

helecta commented on Tracker/CoreData/CoreDataStores/TrackerStore.swift:
```diff
@@ -0,0 +1,180 @@
+        request.predicate = NSPredicate(format: "trackerID = %@", id.uuidString)
+        
+        let trackerID = try! context.execute(request) as! NSAsynchronousFetchResult<NSFetchRequestResult>
+        let id = trackerID.finalResult![0] as! NSManagedObjectID
```
force cast

helecta commented on Tracker/CoreData/CoreDataStores/TrackerStore.swift:
```diff
@@ -0,0 +1,180 @@
+        object.schedule = DaysValue(schedule: newData.date)
+        object.category = categoryCoreData
+        
+        (UIApplication.shared.delegate as! AppDelegate).saveContext()
```
force cast

helecta commented on Tracker/CoreData/CoreDataStores/TrackerStore.swift:
```diff
@@ -0,0 +1,180 @@
+        let idTracker = fetchTrackerID(with: id)
+        let tracker = fetchTracker(with: idTracker)
+        tracker.isPin = !tracker.isPin
+        (UIApplication.shared.delegate as! AppDelegate).saveContext()
```
force cast

helecta commented on Tracker/CoreData/CoreDataStores/TrackerStore.swift:
```diff
@@ -0,0 +1,180 @@
+        request.predicate = NSPredicate(format: "%K == %@",
+                                        #keyPath(TrackerRecordCoreData.recordID),
+                                        id.uuidString)
+        records = try! context.fetch(request)
```
здесь было бы хорошо обработать ошибку

helecta commented on Tracker/CoreData/CoreDataStores/TrackerStore.swift:
```diff
@@ -0,0 +1,180 @@
+        records = try! context.fetch(request)
+        for record in records {
+            context.delete(record)
+            (UIApplication.shared.delegate as! AppDelegate).saveContext()
```
force cast

helecta commented on Tracker/CoreData/CoreDataStores/TrackerStore.swift:
```diff
@@ -0,0 +1,180 @@
+            (UIApplication.shared.delegate as! AppDelegate).saveContext()
+        }
+        context.delete(tracker)
+        (UIApplication.shared.delegate as! AppDelegate).saveContext()
```
force cast

helecta commented on Tracker/CreateTrackers/Category/CategoryViewModel.swift:
```diff
@@ -0,0 +1,47 @@
+    
+    func addNewCategory(_ category: TrackerCategory) {
+        
+        try! categoryStore.addNewCategory(category)
```
здесь было бы хорошо обработать ошибку

helecta commented on Tracker/Resources/en.lproj/Localizable.strings:
```diff
@@ -0,0 +1,59 @@
+"Onboarding1" = " Only track \n what you want";
+"Onboarding2" = " Even if it not \n liters of water and yoga";
```
Ты не сделал в коде локализацию для Onboarding1 и Onboarding2, у тебя в коде "Onboarding 1" и "Onboarding 2"

helecta commented on Tracker/CollectionOfTrackers/TrackerCollectionViewCell.swift:
```diff
@@ -210,3 +243,37 @@ extension TrackerCollectionViewCell {
+        let menuConfig = UIContextMenuConfiguration(actionProvider:  { _ in
+            
+            let fixAction = UIAction(
+                title: self.isPin ? NSLocalizedString("Unpin", comment: "") : NSLocalizedString("Pin", comment: "")
```
Довольно часто пишут для локализации расширение такого типа:
```
extension String {
    
    /// Из какого файла брать строку
    enum Source: String {
        /// Общий файл локализации для всех таргетов
        case common = "LocalizableCommon"
        /// Файл локализации для таргета
        case target = "Localizable"

    }
    
    /**
     Получить локализацию для строки
     - parameter sourceType: источник строки, по умолчанию .common, т.е из общего LocalizableCommon.string
     - parameter comment: комментарий, по умолчанию ""
     - returns: локализованный текст
     */
    func localized(
        from source: Source = .common,
        _ bundle: Bundle = Bundle.main,
        _ comment: String = "") -> String {
        var notFoundValue = self
        // на всякий случай пытаемся найти значение в общем файле, вернем его если не будет найдено в source
        if source != .common {
            notFoundValue = NSLocalizedString(
                self,
                tableName: Source.common.rawValue,
                bundle: bundle,
                comment: comment)
        }
        return NSLocalizedString(
            self,
            tableName: source.rawValue,
            bundle: bundle,
            value: notFoundValue,
            comment: comment)
    }
    
    /**
      Метод возвращает локализованную строку со склонением подходящим для числительного
     - parameter number: число для которого будет подбираться склонение
     - parameter sourceType: источник строки, по умолчанию .common, т.е из общего LocalizableCommon
     - returns: локализованный текст
     */
    func localizedPlural(for number: Int, from source: Source = .common) -> String {
        // TODO: - пока нет локализации для stringsdict, чтобы правильно определять число задаем локаль русскую
        let locale = Locale(identifier: "ru_RU")
        return String(format: self.localized(from: source), locale: locale, arguments: [number])
    }
}
```

Тогда можно в коде так локализацию делать:
```
"confirm_delete_cat".localized()
```
Что значительно короче)

Peppa-Akk replied:
Классное расширение. Спасибо!

helecta commented on Tracker/SplashViewController/TabBar/TrackersViewController/TrackersViewController.swift:
```diff
@@ -0,0 +1,746 @@
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        
+        analyticService.report(event: "Opened TrackersViewController", params: ["event": "open", "screen": "Main"])
```
Подразумевалось что event не в параметрах отправляется, а как name в этом методе:
```
  AppMetrica.reportEvent(name: event, parameters: params, onFailure: { error in
            print("REPORT ERROR: %@", error.localizedDescription)
        })
```
Т.е тут должно быть так: `analyticService.report(event: "open", params: ["screen": "Main"])`

helecta commented on Tracker/SplashViewController/TabBar/TrackersViewController/TrackersViewController.swift:
```diff
@@ -0,0 +1,746 @@
+    }
+}
+
+extension Date {
```
Обрати внимание, что у тебя очень раздулся контроллер, уже 750 строк. Советую выбрать любимую архитектуру, которая поможет избегать проблемы массивного vc. Например, я люблю MVVM. Простая архитектура, которую в будущем с ростом проекта, можно легко доработать на другую.

Peppa-Akk replied:
Хочу после дипломного проекта провести рефакторинг данного приложения. Мне также очень понравилась архитектура MVVM!

Peppa-Akk committed:
Fixes for sprint 17:
- Black rectangle on dark theme in Statistic;
- Add placeholder plug for Statistic;
- Text on Onboarding;
- Retain cycles;
- Delete force-cast operations.
